client/rest/src/main/java/org/elasticsearch/client/ResponseListener.java

```diff
@@ -19,24 +19,22 @@
 
 package org.elasticsearch.client;
 
-import org.apache.http.StatusLine;
-import org.apache.http.client.methods.CloseableHttpResponse;
-import org.apache.http.message.BasicHttpResponse;
-
-import java.io.IOException;
-
 /**
- * Simple {@link CloseableHttpResponse} impl needed to easily create http responses that are closeable given that
- * org.apache.http.impl.execchain.HttpResponseProxy is not public.
+ * Listener to be provided when calling async performRequest methods provided by {@link RestClient}.
+ * Those methods that do accept a listener will return immediately, execute asynchronously, and notify
+ * the listener whenever the request yielded a response, or failed with an exception.
  */
-class CloseableBasicHttpResponse extends BasicHttpResponse implements CloseableHttpResponse {
+public interface ResponseListener {
 
-    public CloseableBasicHttpResponse(StatusLine statusline) {
-        super(statusline);
-    }
+    /**
+     * Method invoked if the request yielded a successful response
+     */
+    void onSuccess(Response response);
```

Comment:
How very familiar!

I wonder if there is anything we can do for Java 8 users to make this possible to Lambda. It can probably wait though. Something like
`
public static class FailureListener {
  private final Consumer<Exception> onFailure;
  public FailureListener(Consumer<Exception> onFailure) {this.onFailure = onFailure;}
  public ResponseListener listener(Consumer<Response> onSuccess) {
    return new ResponseListener() {
      ...
    };
  }
}
```

Maybe it isn't worth doing at all and instead it should be a thing we think about for the high level client.